### ISSUE TITLE
Lie

### ISSUE THREAD

**Issue by shacklefordjames60-coder** [mfouad/awesome-egypt-opensource#29]
"""
qsnn_full_pipeline_v2.py
Enhanced QSNN hybrid polygraph with ONNX export, QEC sim, and expanded tests.

New features:
- Quantum Error Correction: Repetition code (majority vote) in quantum_forward_cos.
- ONNX Export: Standard ML interchange for deployment.
- Tests: QEC mitigation, ONNX round-trip, QEC-forward integration.

WARNING: Probabilistic; human review required. QEC is simulated—real noise varies.
"""

import os
import io
import base64
import math
import random
from collections import deque

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F

# Optional: ONNX Runtime for load/test (pip install onnxruntime if needed)
try:
    import onnxruntime as ort
    _HAS_ONNXRT = True
except ImportError:
    _HAS_ONNXRT = False

# Other optionals (as before)
try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    _HAS_CRYPTO = True
except ImportError:
    AESGCM = None
    _HAS_CRYPTO = False

try:
    import pennylane as qml
    _HAS_PENNYLANE = True
except ImportError:
    qml = None
    _HAS_PENNYLANE = False

try:
    import matplotlib.pyplot as plt
    _HAS_MPL = True
except ImportError:
    plt = None
    _HAS_MPL = False

# Core QSNN model with QEC
class QSNNPolygraph(nn.Module):
    def __init__(self, input_dim=768, hidden_dim=64, num_qubits=4, dropout_p=0.1, device=None, qec_enabled=False):
        super().__init__()
        self.device = device or (torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu"))
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.num_qubits = num_qubits
        self.qec_enabled = qec_enabled  # Enable QEC sim

        self.embed = nn.Linear(input_dim, hidden_dim)
        self.dropout = nn.Dropout(dropout_p)
        self.theta_params = nn.Parameter(torch.randn(hidden_dim, num_qubits) * 0.01)
        self.readout = nn.Linear(hidden_dim + 1, 1)

        self.to(self.device)

    def quantum_forward_cos(self, h, error_rate=0.01):
        thetas = torch.matmul(h, self.theta_params)
        exp_vals = torch.cos(thetas)  # Base expectations

        if self.qec_enabled:
            # Simulate noise on base
            errors = torch.rand_like(exp_vals) < error_rate
            noisy_exp = torch.where(errors, -exp_vals, exp_vals)

            # Repetition code: 3 measurements per qubit
            measurements = noisy_exp.unsqueeze(2).repeat(1, 1, 3)  # batch x num_qubits x 3
            rep_errors = torch.rand_like(measurements) < error_rate
            measurements = torch.where(rep_errors, -measurements, measurements)

            # Majority vote per qubit
            majority_per_qubit = torch.mode(measurements, dim=2)[0]  # batch x num_qubits
            q_out = majority_per_qubit.mean(dim=1, keepdim=True)
        else:
            q_out = exp_vals.mean(dim=1, keepdim=True)

        return q_out

    def forward(self, x, quantum_forward_fn=None):
        x = x.to(self.device)
        h = F.relu(self.embed(x))
        h = self.dropout(h)
        q_fn = quantum_forward_fn or self.quantum_forward_cos
        q_out = q_fn(h)
        combined = torch.cat([h, q_out], dim=1)
        logit = self.readout(combined).squeeze(1)
        return logit

# ... (Include all other functions from previous version: DummyDataset, enable_dropout, mc_dropout_proba, make_pennylane_quantum_forward, temperature_scale_logits, expected_calibration_error, plot_reliability, encrypt_state_dict_b64, decrypt_state_dict_b64_to_model, export_torchscript, ReplayBuffer, autonomous_refine)

# New: ONNX Export
def export_onnx(model, example_input, out_path="qsnn_onnx.onnx", opset_version=11):
    """
    Export to ONNX. Disables QEC for compat (custom ops not supported).
    """
    model_copy = QSNNPolygraph(model.input_dim, model.hidden_dim, model.num_qubits, device=model.device, qec_enabled=False)
    model_copy.load_state_dict(model.state_dict())
    model_copy.eval()
    torch.onnx.export(
        model_copy,
        example_input.to(model.device),
        out_path,
        export_params=True,
        opset_version=opset_version,
        do_constant_folding=True,
        input_names=['input'],
        output_names=['logit'],
        dynamic_axes={'input': {0: 'batch_size'}, 'logit': {0: 'batch_size'}}
    )
    return out_path

# Updated Demo
def demo_pipeline():
    print("=== QSNN v2 Demo: QEC + ONNX ===")
    model = QSNNPolygraph(qec_enabled=True)
    model.optimizer = optim.AdamW(model.parameters(), lr=1e-3)

    # Quick train (as before)
    ds = DummyDataset(512)
    loader = torch.utils.data.DataLoader(ds, 32, True)
    for inp, lbl in loader:
        logits = model(inp)
        loss = nn.BCEWithLogitsLoss()(logits, lbl.squeeze(1))
        model.optimizer.zero_grad()
        loss.backward()
        model.optimizer.step()
        break
    print(f"[demo] Quick train loss: {loss.item():.4f}")

    # QEC Demo
    h = torch.randn(4, model.hidden_dim)
    q_noqec = QSNNPolygraph(qec_enabled=False).quantum_forward_cos(h, error_rate=0.2)
    q_withqec = model.quantum_forward_cos(h, error_rate=0.2)
    print(f"QEC: No correction mean {q_noqec.mean():.4f} | With correction {q_withqec.mean():.4f}")

    # ONNX Export
    example = torch.randn(1, model.input_dim)
    onnx_path = export_onnx(model, example)
    print(f"[demo] ONNX exported: {onnx_path}")

    # (Rest of demo as before: MC, calib, serialize, refine)

    # Autonomous refine with QEC
    rb = ReplayBuffer(1000)
    # Populate (simplified)
    for _ in range(200):
        inp = torch.randn(model.input_dim)
        lbl = torch.tensor([random.choice([0.,1.])])
        rb.add(inp, lbl)
    refine_msg = autonomous_refine(model, rb, threshold=0.1)
    print(f"[demo] Refine: {refine_msg}")

    print("=== Demo Complete ===")

if __name__ == "__main__":
    demo_pipeline()
